src/Uno.UI.Composition/Composition/BorderVisual.skia.cs

```diff
@@ -0,0 +1,196 @@
+
+	private BorderStateWrapper _borderShapeAndBackgroundState;
+	private CompositionSpriteShape? _backgroundShape;
+	private CompositionSpriteShape? _borderShape;
+	// Note how we don't need to get notified when these properties change since they only get set in OnPropertyChangedCore.
+	private CompositionClip? _backgroundClip;
+	private RectangleClip? _childClipCausedByCornerRadius;
+
+	public BorderStateWrapper BorderShapeAndBackgroundState
+	{
+		private get => _borderShapeAndBackgroundState;
+		set => SetObjectProperty(ref _borderShapeAndBackgroundState, value);
+	}
+
+	// BackgroundShape and BorderShape are NOT added to this.Shapes, which both makes it easier
+	// to reason about (no external tampering) and is also closer to what WinUI does.
+	public CompositionSpriteShape BackgroundShape
+	{
+		get
+		{
+			if (_backgroundShape is null)
+			{
+				// we need this to track get notified on brush updates.
+				SetProperty(ref _backgroundShape, Compositor.CreateSpriteShape());
+#if DEBUG
+				_backgroundShape!.Comment = "#borderBackground";
+#endif
+			}
+			return _backgroundShape!;
+		}
+	}
+
+	public CompositionSpriteShape BorderShape
+	{
+		get
+		{
+			if (_borderShape is null)
+			{
+				// we need this to track get notified on brush updates.
+				SetProperty(ref _borderShape, Compositor.CreateSpriteShape());
+#if DEBUG
+				_borderShape!.Comment = "#borderShape";
+#endif
+			}
+			return _borderShape!;
+		}
+	}
+
+	private protected override void ApplyPostPaintingClipping(in SKCanvas canvas)
+		=> _childClipCausedByCornerRadius?.Apply(canvas, this);
+
+	internal override void Paint(in PaintingSession session)
+	{
+		if (_backgroundShape is { } backgroundShape)
+		{
+			session.Canvas.Save();
+			// it's necessary to clip the background because not all backgrounds are simple rounded rectangles with a solid color.
+			// E.g. effect brushes will draw outside the intended area if they're not clipped.
+			_backgroundClip?.Apply(session.Canvas, this);
+			backgroundShape.Render(in session);
+			session.Canvas.Restore();
+		}
+
+		base.Paint(in session);
+
+		_borderShape?.Render(in session);
+	}
+
+	private protected override void OnPropertyChangedCore(string? propertyName, bool isSubPropertyChange)
+	{
+		// Call base implementation - Visual calls Compositor.InvalidateRender().
+		base.OnPropertyChangedCore(propertyName, isSubPropertyChange);
+
+		if (propertyName is not nameof(BorderShapeAndBackgroundState) && propertyName is not nameof(Size))
+		{
+			return;
+		}
+
+		// Note: To reduce allocations, we don't start with a fresh BorderShape and BackgroundShape, but
+		// we modify their properties. This means that the properties might be outdated and we need to set
+		// all relevant properties every time. Any state being set conditionally needs to be accompanied
+		// by an else branch that resets the state to its default value (see CornerRadiusClip below).
```

Comment:
Even if much better than previous code, I have the feeling all of this is still overkill: why are we aggregating all properties into a "state" that we then try to diff to update only relevant properties. Could we instead add callback directly to dependency properties and update the composition properties accordingly (i.e. `BorderShape` and `BackgroundShape`)? 

We can even imagine to have dedicated path / geometry that will reuse the same `SKRoundRect` and update it only if needed (`Size` or `Thickness` changed).